crates/bevy_ecs/src/system/adapter_system.rs

```diff
@@ -0,0 +1,126 @@
+use super::{ReadOnlySystem, System};
+use crate::world::unsafe_world_cell::UnsafeWorldCell;
+
+pub trait Adapt<S: System>: Send + Sync + 'static {
+    /// The [input](System::In) type for an [`AdapterSystem`].
+    type In;
+    /// The [output](System::Out) type for an [`AdapterSystem`].
+    type Out;
+
+    /// When used in an [`AdapterSystem`], this function customizes how the system
+    /// is run and how its inputs/outputs are adapted.
+    fn adapt(&mut self, input: Self::In, run_system: impl FnOnce(S::In) -> S::Out) -> Self::Out;
+}
+
+#[derive(Clone)]
+/// A system that takes the output of `S` and transforms it by applying `Func` to it.
+pub struct AdapterSystem<Func, S> {
+    func: Func,
+    system: S,
+}
+
+impl<Func, S> AdapterSystem<Func, S> {
+    pub fn new(func: Func, system: S) -> Self {
+        Self { func, system }
+    }
+}
+
+impl<Func, S> System for AdapterSystem<Func, S>
+where
+    Func: Adapt<S>,
+    S: System,
+{
+    type In = Func::In;
+    type Out = Func::Out;
+
+    fn name(&self) -> std::borrow::Cow<'static, str> {
+        self.system.name()
+    }
+
+    fn type_id(&self) -> std::any::TypeId {
```

Comment:
This should probably be `TypeId::of::<Self>`, as this is _not_ a system of type `S`. Using the internal `System::name` also gives me pause, but that feels way less "potentially behavior breaking". And generating a unique name would allocate, which I know would undermine one of your goals for this abstraction.

Reply:
The bit about names is a fair objection. I think that losing the `!` for `NotSystem` could be problematic when reporting errors, since it may not be obvious that a run condition is inverted.

I'll try to think on an API for adapting system names.